src/Proto.Cluster/HashRing.cs

```diff
@@ -0,0 +1,50 @@
+﻿// -----------------------------------------------------------------------
+//   <copyright file="HashRing.cs" company="Asynkron HB">
+//       Copyright (C) 2015-2017 Asynkron HB All rights reserved
+//   </copyright>
+// -----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Proto.Cluster
+{
+    public class HashRing
+    {
+        private const int ReplicaCount = 100;
+        private static readonly HashAlgorithm HashAlgorithm = MD5.Create();
+        private readonly List<Tuple<uint, string>> _ring;
+
+        public HashRing(IEnumerable<string> nodes)
+        {
+            _ring = nodes
+                .SelectMany(n => Enumerable.Range(0, ReplicaCount).Select(i => new
+                {
+                    hashKey = i + n,
+                    node = n
+                }))
+                .Select(a => Tuple.Create(Hash(a.hashKey), a.node))
+                .OrderBy(t => t.Item1)
+                .ToList();
+        }
+
+        public string GetNode(string key)
+        {

```

Comment:
I'd vote for using the C# hashring as the base and port it over to Go.
The current Go impl is less effective